onnx/version_converter/convert.cc

```diff
@@ -0,0 +1,122 @@
+    const OpSetID& initial_version,
+    const OpSetID& target_version) const {
+  const std::string initial_domain = initial_version.domain();
+  const std::string target_domain = target_version.domain();
+
+  ONNX_ASSERTM((initial_domain == "" || initial_domain == "ai.onnx") &&
+    (target_domain == "" || target_domain == "ai.onnx"),
+      "Warning: default onnx version converter can only convert "
+      " between default domain opset versions ('' or 'ai.onnx')\n");
+
+  ONNX_ASSERTM(initial_domain == target_domain,
+    "initial_version and target_version must have the same domains");
+  for (auto it = mp_in.opset_import().begin(); it != mp_in.opset_import()
+      .end(); ++it) {
+    if (it->domain() == initial_version.domain()) {
+      ONNX_ASSERTM(initial_version.version() == it->version(),
+          "initial_version does not reflect current state of model");
+    }
+  }
+
+  std::shared_ptr<Graph> g(ImportModelProto(mp_in));
+  ONNX_ASSERTM(g.get() != nullptr,
+    "Warning: onnx version converter is unable to parse input model. "
+    "(The IR version of the ONNX model may be too old.)");
+
+  // TODO: Move to Inter-Domain Converter
+  // Get initial model versions
+  // std::vector<OpSetID> initial_versions = g->opset_versions_mutable();
+
+  // No conversion necessary if Model has single, equivalent opset version
+  // if (initial_versions.size() == 1 && initial_versions[0].version ==
+  //    target_version.version && initial_versions[0].domain ==
+  //    target_version.domain) {
+  //  return mp_in;
+  // }
+
+  // Check if target_version is valid
+  // TODO: Assert same for initial version
+  ONNX_ASSERTM(target_version.version() >= version_range.first && target_version
+      .version() <= version_range.second,
+    "Warning: invalid target_version (must be between %s and %s",
+    version_range.first, version_range.second);
+  // Compile list of all ops used in the model
+  graph_node_list nodes = g->nodes();
+
+  // Iterate over all versions to target_version for specified
+  int64_t curr_version = initial_version.version();
+  int64_t step;
+  bool up;
+  if (target_version.version() > initial_version.version()) {
+    step = 1;
+    up = true;
+  } else {
+    step = -1;
+    up = false;
+  }
+  // Identify index of this domain in g.opset_versions
+  unsigned int domain_index = 0;
+  for (unsigned int i = 0; i < g->opset_versions_mutable().size(); i++) {
+    if (g->opset_versions_mutable()[i].domain() == "") {
+      domain_index = i;
+    }
+  }
+  while (curr_version != target_version.version()) {
+    debug("curr_version: " + std::to_string(curr_version) + ", next_version: " +
+        std::to_string(curr_version + step));
+    // Iterate through and call adapter returned by adapter_lookup for ops from current_version opset
+    for (Node* op : nodes) {
+      debug(std::string("Finding schema for ") + std::string(op->kind().toString()));
+      const std::string& op_name = op->kind().toString();
```

Comment:
this is a reference to a dead object because `.toString()` returns a value, not a reference.